File: programmers/수식최대화.py
from itertools import permutations
import re


def calculate(exp, pri):
    """
    연산자의 우선 순위에 따라 계산하여 산출된 결과를 반환한다.
    동일한 연산자가 있을 경우, 앞에 위치하는 연산자를 먼저 계산한다.
    exp = [operand, operator, operand ...]
    pri = ('*', '+', '-')

    exp = ['100', '-', '200', '*', '300', '-', '500', '+', '20']
    pri = ('*', '+', '-')

    * 연산자가 1순위이므로 * 연산자를 찾아내 앞에 위치한 200과 300을 * 연산자한다.
    * 연산자가 더 이상 없을 때까지 반복하고 다음 우선 순위로 넘어간다.
    값이 하나가 나올 때까지 반복한다.
    """
    p_list = list(pri[::-1])

    p = p_list.pop()
    while True:
        try:
            i = exp.index(p)
            exp = exp[:i - 1] + [str(eval(
                exp[i - 1] + exp[i] + exp[i + 1]
            ))] + exp[i + 2:]

        except ValueError:
            if len(p_list) == 0:
                break
            p = p_list.pop()

    return int(exp[0])


def solution(expression):
    """
    expression 내에 존재하는 연산자를 고른다.

    + - * => 총 6가지의 경우의 수가 나오므로 우선 순위를 설정해 모든 경우를 고려하여 절댓값이 가장 큰 값을 탐색한다.

    regex를 이용하여 operand와 operator를 분리한다.
    """
    regex = re.compile(r"\d+|-|\+|\*|")
    exp = regex.findall(expression)[:-1]

    # 연산자 우선 순위 고려
    operators = set(filter(lambda x: x in ["+", "-", "*"], exp))
    priorities = list(permutations(operators))

    result = 0
    for priority in priorities:
        result = max(result, abs(calculate(exp, priority)))

    return result


print(solution("100-200*300-500+20"))  # 60420
print(solution("50*6-3*2"))  # 300

File: programmers/PART06/수식최대화.js
// https://programmers.co.kr/learn/courses/30/lessons/67257?language=javascript

/**
 * 분할 정복 문제
 *
 * 연산문자: +, -, *
 * expression = "100-200*300-500+20"
 * 정규표현식을 통해 operands와 operators를 분리한다.
 * [100, -, 200, *, 300, -, 500, +, 20]
 *
 * 수식의 결과값을 반환하는 재귀함수를 작성한다.
 * 탈출조건: 배열의 길이가 1일 때, 결과의 절댓값을 반환한다.
 *
 * 연산자가 3개인 경우, 우선순위의 경우의 수는 6가지가 나온다.
 * 우선순위는 순서를 보장해야하므로 순열을 사용한다.
 *
 * [
 *   [ '-', '*', '+' ],
 *   [ '-', '+', '*' ],
 *   [ '*', '-', '+' ],
 *   [ '*', '+', '-' ],
 *   [ '+', '-', '*' ],
 *   [ '+', '*', '-' ]
 * ]
 *
 * 수식과 각 우선순위로 재귀함수를 호출한다.
 * 탈출조건은 수식배열의 크기가 1이 됐을 때, 모든 연산이 종료됐으므로 해당 배열의 숫자의 절댓값을 최댓값과 비교한다.
 *
 */
function solution(expression) {
  const exp = expression.match(/\d+|\D/g);
  const operators = [...new Set(expression.match(/\D/g))];
  const operate = (a, operator, b) => {
    [a, b] = [Number(a), Number(b)];
    if (operator === "+") return a + b;
    else if (operator === "-") return a - b;
    else if (operator === "*") return a * b;
  };
  let answer = 0;

  const recursive = (expr, priority) => {
    if (expr.length === 1) return (answer = Math.max(answer, Math.abs(expr[0])));
    const operator = priority.pop();
    let index = expr.indexOf(operator);
    while (index !== -1) {
      const result = operate(expr[index - 1], operator, expr[index + 1]);
      expr = [...expr.slice(0, index - 1), result, ...expr.slice(index + 2)];
      index = expr.indexOf(operator);
    }
    recursive(expr, priority);
  };

  for (const priority of permutations(operators)) {
    recursive(exp, priority);
  }

  return answer;
}

function permutations(iterable, r = iterable.length) {
  if (r === 1) return iterable.map((value) => [value]);
  return iterable.reduce((result, value, index, origin) => {
    const rest = [...origin.slice(0, index), ...origin.slice(index + 1)];
    return result.concat(permutations(rest, r - 1).map((p) => [value, ...p]));
  }, []);
}

let expression;

expression = "100-200*300-500+20";
console.log(solution(expression)); // 60420

expression = "50*6-3*2";
console.log(solution(expression)); // 300
